ec/src/hashing/curve_maps/elligator2.rs

```diff
@@ -0,0 +1,270 @@
+use crate::models::twisted_edwards::{MontCurveConfig, TECurveConfig};
+use ark_ff::{Field, One, Zero};
+use core::marker::PhantomData;
+
+use crate::{
+    hashing::{curve_maps::parity, map_to_curve_hasher::MapToCurve, HashToCurveError},
+    models::twisted_edwards::{Affine, Projective},
+};
+
+/// Trait defining the necessary parameters for the Elligator2 hash-to-curve method
+/// for twisted edwards curves form of:
+/// `b * y² = x³ + a * x² + x` from [\[WB2019\]]
+/// according to [\[HSSWW23\]]
+///
+/// - [\[BHKL13\]] <http://dx.doi.org/10.1145/2508859.2516734>
+/// - [\[HSSWW23\]] <https://datatracker.ietf.org/doc/html/rfc9380>
+
+pub trait Elligator2Config: TECurveConfig + MontCurveConfig {
+    /// An element of the base field that is not a square root see \[BHKL13, Section 5\].
+    /// When `BaseField` is a prime field, [\[HSSWW23\]] mandates that `Z` is the
+    /// non-square with lowest absolute value in the `BaseField` when its elements
+    /// are represented as [-(q-1)/2, (q-1)/2]
+    const Z: Self::BaseField;
+}
+
+/// Represents the Elligator2 hash-to-curve map defined by `P`.
+pub struct Elligator2Map<P: TECurveConfig>(PhantomData<fn() -> P>);
+
+impl<P: Elligator2Config> MapToCurve<Projective<P>> for Elligator2Map<P> {
+    /// Checks if `P` represents a valid Elligator2 map.
+    fn check_parameters() -> Result<(), HashToCurveError> {
+        // Verifying that Z is a non-square
+        debug_assert!(
+            !P::Z.legendre().is_qr(),
+            "Z should be a quadratic non-residue for the Elligator2 map"
+        );
+
+        // We assume that the Montgomery curve is correct and  as such we do
+        // not verify the prerequisite for applicability of Elligator2 map
+        Ok(())
+    }
+
+    /// Map an arbitrary base field element `element` to a curve point.
+    fn map_to_curve(element: P::BaseField) -> Result<Affine<P>, HashToCurveError> {
+        // 1. x1 = -(J / K) * inv0(1 + Z * u^2)
+        // 2. If x1 == 0, set x1 = -(J / K)
+        // 3. gx1 = x1^3 + (J / K) * x1^2 + x1 / K^2
+        // 4. x2 = -x1 - (J / K)
+        // 5. gx2 = x2^3 + (J / K) * x2^2 + x2 / K^2
+        // 6. If is_square(gx1), set x = x1, y = sqrt(gx1) with sgn0(y) == 1.
+        // 7. Else set x = x2, y = sqrt(gx2) with sgn0(y) == 0.
+        // 8. s = x * K
+        // 9. t = y * K
+        // 10. return (s, t)
+
+        // ark a is irtf J
+        // ark b is irtf k
+        let j = <P as MontCurveConfig>::COEFF_A;
+        let k = <P as MontCurveConfig>::COEFF_B;
+        let j_on_k = j / k;
```

Comment:
Can this be precomputed in some way?

Reply:
@Pratyush I see you have already brought up issue 1. yourself: https://github.com/arkworks-rs/algebra/pull/627#issuecomment-1520918942

Reply:
I'd be happy to revert the `new` removal; I assumed y'all had come to some consensus on that front internally at Web3.

Reply:
So I think the consensus is to resume the `new` so that we have opportunities for precomputation? I agree that this is a good approach because otherwise we need to hardcode the parameters, but these parameters are not super easy to compute: https://github.com/FindoraNetwork/noah/blob/develop/crypto/src/hashing_to_the_curve/ed25519/sswu.rs#L27

Reply:
No we don't have consensus on Web3 side as we have hard-to-convince anti-OOP worriers here. That is why I asked to see what have been on Arkworks reviewers mind to sign off on removing `new` and the ability of having longer-than-one-hash data for MapToCurve.

The counter argument is:

  "This is NOT how substrate+frame works. Frame pretends the lifetime of your code is each tx."

I doubt the validity of the statement (Frame pallets could have longer lived objects) but for the sake of argument suppose it is true.

1. Should Arkworks remove the ability of ever storing any runtime customization/pre-computation by any MapToCurve in future because a specific part of a specific implementation of a specific protocol is not able to take advantage of that optimization. 

2. Even in cases like Elligator where the pre-computation is rather simple and perhaps it makes sense to be outsourced to the curve implementor, it is even more important to enforce verification of the pre-computation in `new`, rather than to leave at the mercy of the same curve implementor to call `check_parameters`. We could feature-decorate the check in `new` so it only happens in debug runs. In this way it won't take an efficiency toll where the code is unable to keep the map alive for more one hash.

Reply:
Why is there even a discussion?  `const fn` can do this just fine.  Or do we have some required methods which cannot be made cost?

Reply:
Trait methods are not currently callable in a `const` context.

Reply:
Yeah I just looked that up.  oops.

I'd say make `check_parameters` do this
```
if cfg!(debug_assertions) {
    let correct = whatever_computation
    if Self::thing != correct {
        println!("Incorrect parameter should be {}",correct);
    }
}
```

Reply:
Maybe we should first run some benchmarks before we fighting imaginary perfomance foot guns and presumed bottlenecks. This is one-hash (so no keep-alive benefit) before and after merging @burdge 's changes.
```
     Running benches/my_benchmark.rs (~/code/arkworks-algebra/target/release/deps/my_benchmark-1efd55fa18a21b6d)
elligator - with new    time:   [3.1494 µs 3.1551 µs 3.1618 µs]
elligator - no new      time:   [3.9473 µs 3.9711 µs 4.0002 µs]
```
It seems that we have lost 20+% in performance. I'm not sure why and I don't know exactly what they have changed. 

My point is that we had `new` and the option of having a stateful map, it was worknig. There are plenty of other useful stuff we could spend our time on, rather than implementing future-foot-gun prevention measures resulting in 20% loss of our hash perfomance.

Reply:
Merged.